pkg/domain/auth/paseto_auth/paseto_token.go

```diff
@@ -0,0 +1,59 @@
+package paseto_auth
+
+import (
+	"aidanwoods.dev/go-paseto"
+	"errors"
+	"strconv"
+	"time"
+)
+
+var (
+	ErrExpiredToken = errors.New("token has expired")
+)
+
+type PasetoToken struct {
+	SignedKey string
+	PublicHex string
+}
+
+func CreateToken(userID uint64) (PasetoToken, error) {
+	userIDstr := strconv.FormatUint(userID, 10)
+
+	token := paseto.NewToken()
+	token.SetIssuedAt(time.Now())
+	token.SetNotBefore(time.Now())
+	token.SetExpiration(time.Now().Add(2 * time.Hour))
+	token.SetString("user-id", userIDstr)
+	token.SetSubject(userIDstr)
+
+	secretKey := paseto.NewV4AsymmetricSecretKey()
+	public := secretKey.Public()
+
+	signed := token.V4Sign(secretKey, nil)
+	return PasetoToken{SignedKey: signed, PublicHex: public.ExportHex()}, nil
+}
+
+func DecryptToken(signedKey, publicHex string) (string, error) {
+	publicKey, err := paseto.NewV4AsymmetricPublicKeyFromHex(publicHex)
+	if err != nil {
+		return "", err
+	}
+
+	parser := paseto.NewParser()
+	token, err := parser.ParseV4Public(publicKey, signedKey, nil)
+	if err != nil {
+		return "", err
+	}
+	expiration, err := token.GetExpiration()
+	if err != nil {
+		return "", err
+	}
+	if expiration.Before(time.Now()) {
+		return "", ErrExpiredToken
+	}
+	subject, err := token.GetSubject()
+	if err != nil {
+		return "", ErrExpiredToken
+	}
+	return subject, err
+}
```

Comment:
Just took a look at this out of interest 😄 

Something you shouldn't allow is for a user to specify the public key will be used to verify the PASETO. The current construction here would allow someone to produce a valid PASETO using their own key (and set whichever fields they like), and then pass in the corresponding public key (which will validate). Essentially allowing PASETOs to be forged because there isn't a source of truth about which keys are valid.

It's okay to support multiple public keys, or even have a key-ID in the PASETO footer which indicates which key should be used, but you'd need to make sure that the keys themselves aren't something you accept as user input.

I'm thinking I should probably update the documentation at my end to make sure that it is clear that public keys shouldn't be accepted from user input.